libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccount.m

```diff
@@ -97,198 +96,273 @@ - (instancetype)initWithCredentials:(SFOAuthCredentials*) credentials {
-    NSString *userPhotoPath = [[SFDirectoryManager sharedManager] directoryForUser:self type:NSLibraryDirectory components:@[@"mobilesdk", @"photos"]];
+    __block NSString *userPhotoPath = nil;
+    dispatch_sync(_syncQueue, ^{
+        userPhotoPath = [self photoPathInternal];
+    });
+    return userPhotoPath;
+}
+
+- (NSString *)photoPathInternal {
+    NSString *userPhotoPath =  [[SFDirectoryManager sharedManager] directoryForOrg:_credentials.organizationId user:_credentials.userId  community:_communityId?:kDefaultCommunityName type:NSLibraryDirectory components:@[@"mobilesdk", @"photos"]];
     [SFDirectoryManager ensureDirectoryExists:userPhotoPath error:nil];
-    return [userPhotoPath stringByAppendingPathComponent:[SFDirectoryManager safeStringForDiskRepresentation:self.credentials.userId]];
+    return [userPhotoPath stringByAppendingPathComponent:[SFDirectoryManager safeStringForDiskRepresentation:_credentials.userId]];
 }
 
 - (UIImage*)photo {
     if (nil == _photo) {
-        NSString *photoPath = [self photoPath];
-        NSFileManager *manager = [[NSFileManager alloc] init];
-        if ([manager fileExistsAtPath:photoPath]) {
-            _photo = [[UIImage alloc] initWithContentsOfFile:photoPath];
-        }
+        __weak __typeof(self) weakSelf = self;
+        dispatch_sync(_syncQueue, ^{
+            __strong typeof(weakSelf) strongSelf = weakSelf;
+            NSString *photoPath = [strongSelf photoPathInternal];
+            NSFileManager *manager = [[NSFileManager alloc] init];
+            if ([manager fileExistsAtPath:photoPath]) {
+                strongSelf->_photo = [[UIImage alloc] initWithContentsOfFile:photoPath];
+            }
+        });
     }
     return _photo;
 }
 
 - (void)setPhoto:(UIImage *)photo {
-    NSError *error = nil;
-    NSString *photoPath = [self photoPath];
-    if (photoPath) {
-        NSFileManager *fm = [NSFileManager defaultManager];
-        if ([fm fileExistsAtPath:photoPath]) {
-            if (![fm removeItemAtPath:photoPath error:&error]) {
-                [SFSDKCoreLogger e:[self class] format:@"Unable to remove previous photo from disk: %@", error];
+    dispatch_barrier_async(_syncQueue, ^{
+        NSError *error = nil;
+        NSString *photoPath = [self photoPathInternal];
+        if (photoPath) {
+            NSFileManager *fm = [NSFileManager defaultManager];
+            if ([fm fileExistsAtPath:photoPath]) {
+                if (![fm removeItemAtPath:photoPath error:&error]) {
+                    [SFSDKCoreLogger e:[self class] format:@"Unable to remove previous photo from disk: %@", error];
+                }
             }
+            NSData *data = UIImagePNGRepresentation(photo);
+            if (![data writeToFile:photoPath options:NSDataWritingAtomic error:&error]) {
+                [SFSDKCoreLogger e:[self class] format:@"Unable to write photo to disk: %@", error];
+            }
+            [self willChangeValueForKey:@"photo"];
```

Comment:
Should this pattern of `willChange` and `didChange` be changed to a notification pattern like the rest of the SDK?

Reply:
We can certainly  this legacy code to NSNotification.